### PR TITLE
[feature-wip](auto-inc)(step-2) support auto-increment column for duplicate table

### PR DESCRIPTION
**Pull request by bobhan1**
# Proposed changes

Issue Number: #18501 

## Changes summary

1. add `AutoIncrementGenerator` for `OlapTable` to maintain the auto-incremce column meta data.
2. add FrontService `getAutoIncrementRange()` for BE coordinator to query
3. fill `null` literals for auto-increment column in `Scanner` if user don't specify that column in stream load task
4. geenerate and fill the values of auto-increment colunms in `VOlapTableSink::_fill_auto_inc_cols()`
## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/runtime/descriptors.cpp [hunk 1]
```diff
@@ -64,7 +64,8 @@ SlotDescriptor::SlotDescriptor(const TSlotDescriptor& tdesc)
           _field_idx(-1),
           _is_materialized(tdesc.isMaterialized),
           _is_key(tdesc.is_key),
-          _need_materialize(tdesc.need_materialize) {}
+          _need_materialize(tdesc.need_materialize),
+          _is_auto_increment(tdesc.is_auto_increment) {}
```
Comment (dataroaring) [inline be/src/runtime/descriptors.cpp:68]: what happens if an old fe and new be work together?

**Review by dataroaring — CHANGES_REQUESTED** [review]

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

File: be/src/vec/sink/vtablet_sink.cpp [hunk 2]
```diff
@@ -968,6 +977,96 @@ void VNodeChannel::mark_close() {
...
+        request.column_id = _column_id;
+        request.length = length;
+
+        auto rpc_begin = std::chrono::system_clock::now();
+        _rpc_status = ThriftRpcHelper::rpc<FrontendServiceClient>(
+                _master_addr.hostname, _master_addr.port,
+                [&request, &result](FrontendServiceConnection& client) {
+                    client->getAutoIncrementRange(result, request);
+                });
+        auto rpc_end = std::chrono::system_clock::now();
```
Comment (dataroaring) [inline be/src/vec/sink/vtablet_sink.cpp:1050]: there are timers in a stop_watch.h .

**Review by dataroaring — CHANGES_REQUESTED** [review]

File: be/src/vec/sink/vtablet_sink.h [hunk 3]
```diff
@@ -447,6 +449,73 @@ class IndexChannel {
...
+        static FetchAutoIncIDExecutor instance;
+        return &instance;
+    }
+
+    std::unique_ptr<ThreadPool> _pool;
+};
+
+// estimate the number of remaining rows based on the rows read and bytes read
+struct RowsEstimator {
+    static constexpr size_t BATCH_INTERVAL = 5000;
```
[the remark this answers is not in the document]
Comment (bobhan1) [reply be/src/vec/sink/vtablet_sink.h:465]: I used to want to use `RowsEstimator` to estimate the remaining rows based on the total size of the stream load files, but later I found that FE set the parameter of `filesize` to -1 for stream load task to indicate that its a stream. So currently this class is not used.
I think I can make `_prefetch_size`/`_batch_size` and `_low_water_level_mark`/`_batch_size` in `AutoIncIDBuffer` config by user

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

File: be/src/vec/sink/vtablet_sink.cpp [hunk 4]
```diff
@@ -1049,6 +1175,18 @@ Status VOlapTableSink::prepare(RuntimeState* state) {
         return Status::InternalError("unknown destination tuple descriptor");
     }
 
+    for (size_t idx = 0; idx < _output_tuple_desc->slots().size(); idx++) {
+        if (_output_tuple_desc->slots()[idx]->is_auto_increment()) {
+            _auto_inc_col_idx = idx;
+            _auto_inc_id_buffer = GlobalAutoIncBuffers::GetInstance()->get_auto_inc_buffer(
```
Comment (HappenLee) [inline be/src/vec/sink/vtablet_sink.cpp:1181]: why here need a global Auto Buffer？

File: be/src/vec/sink/vtablet_sink.cpp [hunk 5]
```diff
@@ -1171,6 +1309,46 @@ void VOlapTableSink::_open_partition(const VOlapTablePartition* partition) {
...
+    DCHECK(!slot->is_nullable());
+
+    vectorized::ColumnPtr src_column_ptr = block->get_by_position(idx).column;
+    auto src_nested_column_ptr =
+            assert_cast<const vectorized::ColumnNullable&>(*src_column_ptr).get_nested_column_ptr();
+
+    auto dst_column = vectorized::ColumnInt64::create();
+    vectorized::ColumnInt64::Container& dst_values = dst_column->get_data();
+    dst_values.reserve(rows);
+
```
Comment (HappenLee) [inline be/src/vec/sink/vtablet_sink.cpp:1325]: // the code should be DEBUG
```
size_t count = 0;
    for (size_t i = 0; i < rows; i++) {
        if (src_column_ptr->is_null_at(i)) {
            ++count;
        }
    }
```

File: be/src/vec/sink/vtablet_sink.cpp [hunk 6]
```diff
@@ -964,6 +974,122 @@ void VNodeChannel::mark_close() {
...
+FetchAutoIncIDExecutor::FetchAutoIncIDExecutor() {
+    ThreadPoolBuilder("AsyncFetchAutoIncIDExecutor")
+            .set_min_threads(config::auto_inc_fetch_thread_num)
+            .set_max_threads(config::auto_inc_fetch_thread_num)
+            .set_max_queue_size(std::numeric_limits<int>::max())
+            .build(&_pool);
+}
+
+AutoIncIDRequestHandlerExecutor::AutoIncIDRequestHandlerExecutor() {
+    ThreadPoolBuilder("AutoIncIDRequestHandlerExecutor")
```
Comment (HappenLee) [inline be/src/vec/sink/vtablet_sink.cpp:986]: why here need tow thread pool?

File: be/src/vec/sink/vtablet_sink.cpp (hunk 4 above)
Comment (bobhan1) [reply be/src/vec/sink/vtablet_sink.cpp:1181]: There may be multiple `VOlapTableSink`s on one BE writing to the same table concurrently. Sharing a global buffer between these `VOlapTableSink`s can reduce the number of RPCs

File: be/src/vec/sink/vtablet_sink.cpp (hunk 6 above)
Comment (bobhan1) [reply be/src/vec/sink/vtablet_sink.cpp:986]: One for RPC and the other for handling requestss from `VOlapTableSink`. Putting these two kinds of threads in one thread pool may cause deadlock.

File: be/src/vec/sink/vtablet_sink.cpp [hunk 7]
```diff
@@ -1171,6 +1309,46 @@ void VOlapTableSink::_open_partition(const VOlapTablePartition* partition) {
...
+    }
+
+    DCHECK(_request_future.valid());
+
+    // if the pipeline is enabled, this will not be blocked,
+    // if the pipeline is not enabled, this will not be blocked in most occasions
+    RETURN_IF_ERROR(_request_future.get());
+    DCHECK(_auto_inc_id_allocator.total_count >= count);
+    for (size_t i = 0; i < rows; i++) {
+        if (src_column_ptr->is_null_at(i)) {
```
[the remark this answers is not in the document]
Comment (bobhan1) [reply be/src/vec/sink/vtablet_sink.cpp:1340]: OK, I'll extract the null bitmap out and add SIMD code for this process

File: gensrc/thrift/FrontendService.thrift [hunk 8]
```diff
@@ -1040,6 +1040,20 @@ struct TGetMasterTokenResult {
     2: optional string token
 }
 
+struct TAutoIncrementRangeRequest {
+    1: required i64 db_id;
+    2: required i64 table_id;
+    3: required i64 column_id;
+    4: required i64 length
+    5: optional i64 lower_bound // if set, values in result range must larger than `lower_bound`
+}
```
Comment (dataroaring) [inline gensrc/thrift/FrontendService.thrift:1049]: please donot use required.

File: fe/fe-core/src/main/java/org/apache/doris/catalog/OlapTable.java [hunk 9]
```diff
@@ -2121,4 +2141,18 @@ public boolean isDupKeysOrMergeOnWrite() {
                 || (getKeysType() == KeysType.UNIQUE_KEYS
                 && getEnableUniqueKeyMergeOnWrite());
     }
+
+    public void initAutoIncrentGenerator(long dbId) {
```
Comment (dataroaring) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/OlapTable.java:2145]: dbId is not used.

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 39.37 seconds
 stream load tsv:          457 seconds loaded 74807831229 Bytes, about 156 MB/s
 stream load json:         22 seconds loaded 2358488459 Bytes, about 102 MB/s
 stream load orc:          57 seconds loaded 1101869774 Bytes, about 18 MB/s
 stream load parquet:          29 seconds loaded 861443392 Bytes, about 28 MB/s
 insert into select:          69.9 seconds inserted 10000000 Rows, about 143K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230630152035_clickbench_pr_170615.html

**Comment by bobhan1** [comment]
run build all

**Comment by bobhan1** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 42.06 seconds
 stream load tsv:          455 seconds loaded 74807831229 Bytes, about 156 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          57 seconds loaded 1101869774 Bytes, about 18 MB/s
 stream load parquet:          28 seconds loaded 861443392 Bytes, about 29 MB/s
 insert into select:          69.1 seconds inserted 10000000 Rows, about 144K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230704050456_clickbench_pr_171769.html

**Review by zhannngchen — DISMISSED** [review]
LGTM

**Comment by bobhan1** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 50.98 seconds
 stream load tsv:          502 seconds loaded 74807831229 Bytes, about 142 MB/s
 stream load json:         18 seconds loaded 2358488459 Bytes, about 124 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.1 seconds inserted 10000000 Rows, about 343K ops/s
 storage size: 17161012474 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230713123724_clickbench_pr_177785.html

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 53.25 seconds
 stream load tsv:          501 seconds loaded 74807831229 Bytes, about 142 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          29.3 seconds inserted 10000000 Rows, about 341K ops/s
 storage size: 17166124316 Bytes

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run ckb

**Comment by bobhan1** [comment]
run buildall

**Review by kaijchen — CHANGES_REQUESTED** [review]

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run arm

**Comment by bobhan1** [comment]
run feut

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 52.04 seconds
 stream load tsv:          509 seconds loaded 74807831229 Bytes, about 140 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          32 seconds loaded 861443392 Bytes, about 25 MB/s
 insert into select:          29.3 seconds inserted 10000000 Rows, about 341K ops/s
 storage size: 17161007208 Bytes

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 44.99 seconds
 stream load tsv:          505 seconds loaded 74807831229 Bytes, about 141 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          32 seconds loaded 861443392 Bytes, about 25 MB/s
 insert into select:          30.4 seconds inserted 10000000 Rows, about 328K ops/s
 storage size: 17170174174 Bytes

**Review by dataroaring — APPROVED** [review]
LGTM

**Review by zhannngchen — APPROVED** [review]
LGTM

**Pull request merged**